extensions/extensions-negotiation.md

```diff
@@ -0,0 +1,112 @@
+# Stratum V2 Extension: Extensions Negotiation
+
+## 0. Abstract
+
+This document defines a Stratum V2 extension to negotiate support for other protocol extensions between clients and servers.
+
+The mechanism allows clients to request support for a list of extensions immediately after the `SetupConnection` message exchange. The server responds with either:
+- `RequestExtensions.Success`, listing the supported extensions.
+- `RequestExtensions.Error`, listing the unsupported extensions.
+
+This negotiation ensures that both parties can establish a common set of features before exchanging further protocol messages.
+
+Terms like "MUST," "MUST NOT," "REQUIRED," etc., follow RFC2119 standards.
+
+---
+
+## 1. Overview
+
+### 1.1 Negotiating Extensions
+
+After the successful `SetupConnection` exchange, clients MUST send a `RequestExtensions` message to indicate the extensions they wish to use. The server responds with:
+- `RequestExtensions.Success`, confirming which extensions are supported.
+- `RequestExtensions.Error`, identifying unsupported extensions.
+
+Clients MUST NOT use any features from extensions that are not confirmed as supported by the server.
+
+#### Message Exchange Example
+
+1. **Connection Setup**:
+    ```
+    Client --- SetupConnection (connection details) ---> Server
+    Client <--- SetupConnection.Success (connection accepted) ---- Server
+    ```
+
+2. **Extension Negotiation**:
+    ```
+    Client --- RequestExtensions (list of requested U16) ---> Server
+    ```
+
+3. **Server Response**:
+    - If successful:
+      ```
+      Client <--- RequestExtensions.Success (list of supported U16) ---- Server
+      ```
+    - If an error occurs:
+      ```
+      Client <--- RequestExtensions.Error (list of unsupported U16) ---- Server
+      ```
+
+---
+
+## 2. Messages Defined by This Extension
+
+### `RequestExtensions` (Client -> Server)
+
+| Field Name           | Data Type    | Description                                   |
+|----------------------|--------------|-----------------------------------------------|
+| request_id           | U16          | Unique identifier for pairing the response.  |
+| requested_extensions | SEQ0_64K[U16]| List of requested extension identifiers.     |
+
+### `RequestExtensions.Success` (Server -> Client)
+
+| Field Name           | Data Type    | Description                                   |
+|----------------------|--------------|-----------------------------------------------|
+| request_id           | U16          | Unique identifier for pairing the response.  |
+| supported_extensions | SEQ0_64K[U16]| List of supported extension identifiers.     |
+
+### `RequestExtensions.Error` (Server -> Client)
```

Comment:
What happens if upstream returns `RequestExtensionsError`? are they expected to drop the connection?